// apollo-angular/esm2015/testing/operation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ApolloError } from 'apollo-client';
/** @type {?} */
const isApolloError = (err) => err && err.hasOwnProperty('graphQLErrors');
const ɵ0 = isApolloError;
export class TestOperation {
    /**
     * @param {?} operation
     * @param {?} observer
     */
    constructor(operation, observer) {
        this.operation = operation;
        this.observer = observer;
    }
    /**
     * @param {?} result
     * @return {?}
     */
    flush(result) {
        if (isApolloError(result)) {
            this.observer.error(result);
        }
        else {
            this.observer.next((/** @type {?} */ (result)));
            this.observer.complete();
        }
    }
    /**
     * @param {?} error
     * @return {?}
     */
    networkError(error) {
        /** @type {?} */
        const apolloError = new ApolloError({
            networkError: error,
        });
        this.flush(apolloError);
    }
    /**
     * @param {?} errors
     * @return {?}
     */
    graphqlErrors(errors) {
        this.flush({
            errors,
        });
    }
}
if (false) {
    /** @type {?} */
    TestOperation.prototype.operation;
    /** @type {?} */
    TestOperation.prototype.observer;
}
export { ɵ0 };
