// apollo-angular/fesm2015/ng.apollo.testing.js
import { ApolloError } from 'apollo-client';
import { Injectable, NgModule, InjectionToken, Inject, Optional } from '@angular/core';
import { Observable, ApolloLink } from 'apollo-link';
import { print } from 'graphql/language/printer';
import { ApolloModule, Apollo } from 'apollo-angular';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloCache } from 'apollo-cache';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Controller to be injected into tests, that allows for mocking and flushing
 * of operations.
 *
 *
 * @abstract
 */
class ApolloTestingController {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const isApolloError = (err) => err && err.hasOwnProperty('graphQLErrors');
class TestOperation {
    /**
     * @param {?} operation
     * @param {?} observer
     */
    constructor(operation, observer) {
        this.operation = operation;
        this.observer = observer;
    }
    /**
     * @param {?} result
     * @return {?}
     */
    flush(result) {
        if (isApolloError(result)) {
            this.observer.error(result);
        }
        else {
            this.observer.next((/** @type {?} */ (result)));
            this.observer.complete();
        }
    }
    /**
     * @param {?} error
     * @return {?}
     */
    networkError(error) {
        /** @type {?} */
        const apolloError = new ApolloError({
            networkError: error,
        });
        this.flush(apolloError);
    }
    /**
     * @param {?} errors
     * @return {?}
     */
    graphqlErrors(errors) {
        this.flush({
            errors,
        });
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * A testing backend for `Apollo`.
 *
 * `ApolloTestingBackend` works by keeping a list of all open operations.
 * As operations come in, they're added to the list. Users can assert that specific
 * operations were made and then flush them. In the end, a verify() method asserts
 * that no unexpected operations were made.
 */
class ApolloTestingBackend {
    constructor() {
        /**
         * List of pending operations which have not yet been expected.
         */
        this.open = [];
    }
    /**
     * Handle an incoming operation by queueing it in the list of open operations.
     * @param {?} op
     * @return {?}
     */
    handle(op) {
        return new Observable((observer) => {
            /** @type {?} */
            const testOp = new TestOperation(op, observer);
            this.open.push(testOp);
        });
    }
    /**
     * Helper function to search for operations in the list of open operations.
     * @param {?} match
     * @return {?}
     */
    _match(match) {
        if (typeof match === 'string') {
            return this.open.filter(testOp => testOp.operation.operationName === match);
        }
        else if (typeof match === 'function') {
            return this.open.filter(testOp => match(testOp.operation));
        }
        else {
            if (this.isDocumentNode(match)) {
                return this.open.filter(testOp => print(testOp.operation.query) === print(match));
            }
            return this.open.filter(testOp => this.matchOp(match, testOp));
        }
    }
    /**
     * @param {?} match
     * @param {?} testOp
     * @return {?}
     */
    matchOp(match, testOp) {
        /** @type {?} */
        const variables = JSON.stringify(match.variables);
        /** @type {?} */
        const extensions = JSON.stringify(match.extensions);
        /** @type {?} */
        const sameName = this.compare(match.operationName, testOp.operation.operationName);
        /** @type {?} */
        const sameVariables = this.compare(variables, testOp.operation.variables);
        /** @type {?} */
        const sameQuery = print(testOp.operation.query) === print(match.query);
        /** @type {?} */
        const sameExtensions = this.compare(extensions, testOp.operation.extensions);
        return sameName && sameVariables && sameQuery && sameExtensions;
    }
    /**
     * @param {?=} expected
     * @param {?=} value
     * @return {?}
     */
    compare(expected, value) {
        /** @type {?} */
        const prepare = (val) => typeof val === 'string' ? val : JSON.stringify(val);
        /** @type {?} */
        const received = prepare(value);
        return !expected || received === expected;
    }
    /**
     * Search for operations in the list of open operations, and return all that match
     * without asserting anything about the number of matches.
     * @param {?} match
     * @return {?}
     */
    match(match) {
        /** @type {?} */
        const results = this._match(match);
        results.forEach(result => {
            /** @type {?} */
            const index = this.open.indexOf(result);
            if (index !== -1) {
                this.open.splice(index, 1);
            }
        });
        return results;
    }
    /**
     * Expect that a single outstanding request matches the given matcher, and return
     * it.
     *
     * operations returned through this API will no longer be in the list of open operations,
     * and thus will not match twice.
     * @param {?} match
     * @param {?=} description
     * @return {?}
     */
    expectOne(match, description) {
        description = description || this.descriptionFromMatcher(match);
        /** @type {?} */
        const matches = this.match(match);
        if (matches.length > 1) {
            throw new Error(`Expected one matching operation for criteria "${description}", found ${matches.length} operations.`);
        }
        if (matches.length === 0) {
            throw new Error(`Expected one matching operation for criteria "${description}", found none.`);
        }
        return matches[0];
    }
    /**
     * Expect that no outstanding operations match the given matcher, and throw an error
     * if any do.
     * @param {?} match
     * @param {?=} description
     * @return {?}
     */
    expectNone(match, description) {
        description = description || this.descriptionFromMatcher(match);
        /** @type {?} */
        const matches = this.match(match);
        if (matches.length > 0) {
            throw new Error(`Expected zero matching operations for criteria "${description}", found ${matches.length}.`);
        }
    }
    /**
     * Validate that there are no outstanding operations.
     * @return {?}
     */
    verify() {
        /** @type {?} */
        const open = this.open;
        if (open.length > 0) {
            // Show the methods and URLs of open operations in the error, for convenience.
            /** @type {?} */
            const operations = open
                .map(testOp => testOp.operation.operationName)
                .join(', ');
            throw new Error(`Expected no open operations, found ${open.length}: ${operations}`);
        }
    }
    /**
     * @param {?} docOrOp
     * @return {?}
     */
    isDocumentNode(docOrOp) {
        return !((/** @type {?} */ (docOrOp))).operationName;
    }
    /**
     * @param {?} matcher
     * @return {?}
     */
    descriptionFromMatcher(matcher) {
        if (typeof matcher === 'string') {
            return `Match operationName: ${matcher}`;
        }
        else if (typeof matcher === 'object') {
            if (this.isDocumentNode(matcher)) {
                return `Match DocumentNode`;
            }
            /** @type {?} */
            const name = matcher.operationName || '(any)';
            /** @type {?} */
            const variables = JSON.stringify(matcher.variables) || '(any)';
            return `Match operation: ${name}, variables: ${variables}`;
        }
        else {
            return `Match by function: ${matcher.name}`;
        }
    }
}
ApolloTestingBackend.decorators = [
    { type: Injectable }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const APOLLO_TESTING_CACHE = new InjectionToken('apollo-angular/testing cache');
class ApolloTestingModule {
    /**
     * @param {?} apollo
     * @param {?} backend
     * @param {?=} cache
     */
    constructor(apollo, backend, cache) {
        apollo.create({
            link: new ApolloLink(operation => backend.handle(operation)),
            cache: cache ||
                new InMemoryCache({
                    addTypename: false,
                }),
        });
    }
}
ApolloTestingModule.decorators = [
    { type: NgModule, args: [{
                imports: [ApolloModule],
                providers: [
                    ApolloTestingBackend,
                    { provide: ApolloTestingController, useExisting: ApolloTestingBackend },
                ],
            },] }
];
/** @nocollapse */
ApolloTestingModule.ctorParameters = () => [
    { type: Apollo },
    { type: ApolloTestingBackend },
    { type: ApolloCache, decorators: [{ type: Optional }, { type: Inject, args: [APOLLO_TESTING_CACHE,] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { ApolloTestingController, ApolloTestingModule, APOLLO_TESTING_CACHE, TestOperation, ApolloTestingBackend as ɵa };
