// apollo-angular/esm5/testing/operation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ApolloError } from 'apollo-client';
/** @type {?} */
var isApolloError = function (err) {
    return err && err.hasOwnProperty('graphQLErrors');
};
var ɵ0 = isApolloError;
var TestOperation = /** @class */ (function () {
    function TestOperation(operation, observer) {
        this.operation = operation;
        this.observer = observer;
    }
    /**
     * @param {?} result
     * @return {?}
     */
    TestOperation.prototype.flush = /**
     * @param {?} result
     * @return {?}
     */
    function (result) {
        if (isApolloError(result)) {
            this.observer.error(result);
        }
        else {
            this.observer.next((/** @type {?} */ (result)));
            this.observer.complete();
        }
    };
    /**
     * @param {?} error
     * @return {?}
     */
    TestOperation.prototype.networkError = /**
     * @param {?} error
     * @return {?}
     */
    function (error) {
        /** @type {?} */
        var apolloError = new ApolloError({
            networkError: error,
        });
        this.flush(apolloError);
    };
    /**
     * @param {?} errors
     * @return {?}
     */
    TestOperation.prototype.graphqlErrors = /**
     * @param {?} errors
     * @return {?}
     */
    function (errors) {
        this.flush({
            errors: errors,
        });
    };
    return TestOperation;
}());
export { TestOperation };
if (false) {
    /** @type {?} */
    TestOperation.prototype.operation;
    /** @type {?} */
    TestOperation.prototype.observer;
}
export { ɵ0 };
